zebra-chain/src/parallel/tree.rs

```diff
@@ -160,36 +161,96 @@ impl NoteCommitmentTrees {
     }
 
     /// Update the sapling note commitment tree.
+    #[allow(clippy::unwrap_in_result)]
     fn update_sapling_note_commitment_tree(
         mut sapling: Arc<sapling::tree::NoteCommitmentTree>,
         sapling_note_commitments: Vec<sapling::tree::NoteCommitmentUpdate>,
-    ) -> Result<Arc<sapling::tree::NoteCommitmentTree>, NoteCommitmentTreeError> {
+    ) -> Result<
+        (
+            Arc<sapling::tree::NoteCommitmentTree>,
+            Option<(u16, sapling::tree::Node)>,
+        ),
+        NoteCommitmentTreeError,
+    > {
         let sapling_nct = Arc::make_mut(&mut sapling);
+        let next_complete_subtree_note_index = sapling_nct.next_complete_subtree_note_index();
+
+        // It is impossible for blocks to contain more than one level 16 sapling root:
+        // > [NU5 onward] nSpendsSapling, nOutputsSapling, and nActionsOrchard MUST all be less than 2^16.
+        // <https://zips.z.cash/protocol/protocol.pdf#txnconsensus>
+        //
+        // Before NU5, this limit holds due to the minimum size of Sapling outputs (948 bytes)
+        // and the maximum size of a block:
+        // > The size of a block MUST be less than or equal to 2000000 bytes.
+        // <https://zips.z.cash/protocol/protocol.pdf#blockheader>
+        // <https://zips.z.cash/protocol/protocol.pdf#txnencoding>
+        let mut subtree_root = None;
 
-        for sapling_note_commitment in sapling_note_commitments {
-            sapling_nct.append(sapling_note_commitment)?;
+        if sapling_note_commitments.len() > next_complete_subtree_note_index {
+            for (note_idx, sapling_note_commitment) in
+                sapling_note_commitments.into_iter().enumerate()
+            {
+                sapling_nct.append(sapling_note_commitment)?;
+                if note_idx == next_complete_subtree_note_index {
+                    let node = sapling_nct
+                        .subtree_root()
+                        .expect("should be frontier after appending a note");
+                    subtree_root = Some((sapling_nct.subtree_index(), node));
+                }
+            }
+        } else {
+            for sapling_note_commitment in sapling_note_commitments {
+                sapling_nct.append(sapling_note_commitment)?;
+            }
```

Comment:
I struggled to work out if this code was correct, and what exactly it does. I found it hard to check the method names, bitwise arithmetic, iterator indexes, loop conditions, and off by one errors at the same time.

But I think it might be equivalent to checking for a complete subtree after every node?

```suggestion
        for sapling_note_commitment in sapling_note_commitments {
            sapling_nct.append(sapling_note_commitment)?;
            if sapling_nct.is_complete_subtree() {
                let subtree_root_node = sapling_nct
                    .subtree_root()
                    .expect("should be frontier after appending a note");
                subtree_root = Some((sapling_nct.subtree_index(), subtree_root_node));
            }
```

Reply:
> But I think it might be equivalent to checking for a complete subtree after every node?

Yep.